### PR TITLE
[HigherOrderOp] don't mannually set input for cond

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111611
* #111610

We set mannualy_set_graph_inputs to False for CondHigherOrder. After that, it became necessary to deduplicate the inputs.  We'll add pytree tests in the follow-up pr.

Test Plan:
existing tests.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @zou3519

### CODE REVIEW

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -1247,10 +1247,15 @@ def false_fn(x):
...
+            if (
+                node.op == "call_function"
+                and node.target == torch.ops.higher_order.cond
+            ):
                 _, _, _, operands = node.args
-                # Each branch takes 5 inputs (x, true_buffer, true_z, false_buffer, false_z)
-                self.assertEqual(len(operands), 5)
+                # Each branch takes 4 inputs (x, z, buffer_true_branch, buffer_false_branch)
+                # TODO: we should be able to de-duplicate the buffer accessed from two branches so that
+                # operands become (x, z, buffer)
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:1257]: Why isn't the buffer deduplicated?

File: test/dynamo/test_export.py [hunk 2]
```diff
@@ -3637,34 +3637,40 @@ def forward(self, pred, x):
     ones_3 = torch.ones(6, 4)
     cond_true_0 = self.cond_true_0
     cond_false_0 = self.cond_false_0
-    cond = torch.ops.higher_order.cond(arg0, cond_true_0, cond_false_0, [arg1, ones, ones_1, ones_3, ones, ones_1, ones_2]);  arg0 = cond_true_0 = cond_false_0 = arg1 = ones = ones_1 = ones_3 = ones_2 = None
+    cond = torch.ops.higher_order.cond(arg0, cond_true_0, cond_false_0, [ones, ones_1, arg1, ones_3, ones_2]);  arg0 = cond_true_0 = cond_false_0 = ones = ones_1 = arg1 = ones_3 = ones_2 = None
```
Comment (zou3519) [inline test/dynamo/test_export.py:3640]: cc @voznesenskym why don't these variable names get named better? (e.g. ones_1, ones_2, ones_3) would ideally be (a, b, c)

**Review by zou3519 — APPROVED** [review]
I dont know what's going on with the buffers, but this LGTM and is an improvement over the current state

File: test/dynamo/test_export.py (hunk 2 above)
Comment (voznesenskym) [reply test/dynamo/test_export.py:3640]: If you run with `TORCH_LOGS=+dynamo` and breakpoint in `lazy_format_graph_code`, called from `compile_and_call_fx_graph`, the graph looks right:

```
def forward(self, L_pred_ : torch.Tensor, L_x_ : torch.Tensor):
    l_pred_ = L_pred_
    l_x_ = L_x_
    a = torch.ones(6, 4)
    b = torch.ones(6, 4)
    c = torch.ones(6, 4)
    d = torch.ones(6, 4)
    cond_true_0 = self.cond_true_0
    cond_false_0 = self.cond_false_0
    cond = torch.ops.higher_order.cond(l_pred_, cond_true_0, cond_false_0, [a, b, l_x_, d, c]);  l_pred_ = cond_true_0 = cond_false_0 = l_x_ = None
    return (d, c, b, a, cond)
```

Setting a breakpoint in the export interceding compiler also shows the gm as correctly printing.

Printing after rewrite_signature shows it as changing to a worse version. I can fix this rq.

Comment (voznesenskym) [reply test/dynamo/test_export.py:3640]: https://github.com/pytorch/pytorch/pull/111947

File: test/dynamo/test_higher_order_ops.py (hunk 1 above)
Comment (zou3519) [reply test/dynamo/test_higher_order_ops.py:1257]: thanks for the explanation. Let's figure out how to do this as a follow-up

**Pull request closed without merging**